hal/spi_api.h

```diff
@@ -19,220 +18,259 @@
+     *  testing.
+     */
+    uint32_t    maximum_frequency;
+    /** Each bit represents the corresponding word length. lsb => 1bit, msb => 32bit. */
+    uint32_t    word_length;
+    bool        support_slave_mode; /**< If true, the device can handle SPI slave mode. */
+} spi_capabilities_t;
 
-/** Initialize the SPI peripheral
+/**
+ * This structure groups all required data to handle a SPI transaction.
  *
- * Configures the pins used by SPI, sets a default format and frequency, and enables the peripheral
- * @param[out] obj  The SPI object to initialize
- * @param[in]  mosi The pin to use for MOSI
- * @param[in]  miso The pin to use for MISO
- * @param[in]  sclk The pin to use for SCLK
- * @param[in]  ssel The pin to use for SSEL
+ * A symbol might be bigger than a byte. In such case symbols are read/written following the
+ * platform's endianness.
  */
-void spi_init(spi_t *obj, PinName mosi, PinName miso, PinName sclk, PinName ssel);
+typedef struct {
+    const uint8_t *tx; /**< A buffer containing the data to be sent. */
+    uint32_t tx_count; /**< The number of symbol to send. */
+    uint8_t *rx; /**< A buffer to store the received data. */
+    uint32_t rx_count; /**< The number of symbol to read. */
+} spi_transfer_args_t;
 
-/** Release a SPI object
- *
- * TODO: spi_free is currently unimplemented
- * This will require reference counting at the C++ level to be safe
- *
- * Return the pins owned by the SPI object to their reset state
- * Disable the SPI peripheral
- * Disable the SPI clock
- * @param[in] obj The SPI object to deinitialize
- */
-void spi_free(spi_t *obj);
+#ifdef __cplusplus
+extern "C" {
+#endif
 
-/** Configure the SPI format
+/**
+ * Obtains the capabilities of a SPI channel.
  *
- * Set the number of bits per frame, configure clock polarity and phase, shift order and master/slave mode.
- * The default bit order is MSB.
- * @param[in,out] obj   The SPI object to configure
- * @param[in]     bits  The number of bits per frame
- * @param[in]     mode  The SPI mode (clock polarity, phase, and shift direction)
- * @param[in]     slave Zero for master mode or non-zero for slave mode
- */
-void spi_format(spi_t *obj, int bits, int mode, int slave);
-
-/** Set the SPI baud rate
+ * - returns a constant pointer to a constant structure describing the capabilities of the
+ *   related peripheral regardless of whether the device is in use or not.
  *
- * Actual frequency may differ from the desired frequency due to available dividers and bus clock
- * Configures the SPI peripheral's baud rate
- * @param[in,out] obj The SPI object to configure
- * @param[in]     hz  The baud rate in Hz
+ * @param[in]   pins    spi_pins_t structure identifying a SPI channel.
+ * @return A constant pointer to a constant structure describing the capabilities of the device.
  */
-void spi_frequency(spi_t *obj, int hz);
+const spi_capabilities_t *const spi_get_capabilities(const spi_pins_t *pins);
```

Comment:
Capabilities should be per device and not per pin.

Reply:
Why this function returns address of the `spi_capabilities_t` structure? In such scenario `spi_get_capabilities()` function would have to allocate special memory for the result on heap.
I think that the better option is to take address of the `spi_capabilities_t` structure as a parameter and fills the structure with proper data.

Reply:
This is because the capabilities are not dynamic and should be defined  as static structures.
```c
const spi_capabilities_t my_capabilities_for_spi1 = {
[...]
};
const spi_capabilities_t my_capabilities_for_spi2= {
[...]
};
```

This method should identify the correct structure and return a pointer to this flash stored object.